frontend/cypress/support/commands.ts

```diff
@@ -41,8 +41,98 @@
 //
 // -- This will overwrite an existing command --
 // Cypress.Commands.overwrite("visit", (originalFn, url, options) => { ... })
-// @ts-expect-error there is no type definition for this package yet
-// eslint-disable-next-line no-restricted-imports
-import * as cypressImageDiffCommand from 'cypress-image-diff-js/dist/command';
 
-cypressImageDiffCommand();
+const performedActionsStack: any[] = [];
+
+export function dragToMap(
+    elementSelector: string,
+    offset?: { x: number; y: number }
+) {
+    cy.log(`dragging ${elementSelector} to the map`);
+    const mapSelector = '[data-cy=openLayersContainer]';
+
+    cy.get(elementSelector).first().trigger('mousedown');
+
+    return (
+        cy
+            .get(mapSelector)
+            /**
+             * We have to force the click here, as the img
+             * of the element being dragged is in front of the map
+             **/
+            .trigger('mousemove', { force: true, ...offset })
+            .click({ force: true })
+    );
+}
+
+export function storeState() {
+    return cy
+        .window()
+        .its('store')
+        .then((store: any) => store.source._value.application);
+}
+
+export function socket() {
+    return cy.window().its('socket');
+}
+
+export function registerSocketListener() {
+    performedActionsStack.length = 0;
+    cy.socket().then((s) =>
+        s.on('performAction', (action: any) =>
+            performedActionsStack.push(action)
+        )
```

Comment:
I think this would break in the future.

At the moment, a performed action is sent to all clients in an exercise, including the client that initially proposed it. The proposing client gets additionally a positive response to its proposal-request. This complicates the `OptimisticActionHandler` as we have to identify the `performAction`, that corresponds to our successful proposal (which means we have to compare actions by value). This also leads to a special case where the [`optimisticUpdateHandler` behaves unitntuivively](https://github.dev/hpi-sam/digital-fuesim-manv/blob/b458be325d600ea5f5456b5c743695b972c0a553/frontend/src/app/core/optimistic-action-handler.spec.ts#L290-L291).

This could be fixed by excluding the proposing client from the corresponding `performAction` already on the server. (I initially didn't want to do this, as I thought we would use Rooms from `socket.io`, which have a `broadcast`-method, which I thought could be more performant - but we don't use these rooms, and the performance benefit would be negligible/non-existent).


Alternatives:
* Add these actions directly to the window-object in the `proposeAction` method in the `ExerciseService`
* Maybe use the Ngrx-Store and its reducers somehow?

Reply:
Thanks for the hint. Listening to both proposed and performed actions is probably a good choice, as we are going to need them for later testing.

I'll also take a look how one could integrate the ngrx store properly into cypress.